// mx-react-components/dist/components/DatePickerFullScreen.js
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react');
var Radium = require('radium');
var moment = require('moment');

var Icon = require('./Icon');

var StyleConstants = require('../constants/Style');

var DatePickerFullScreen = (function (_React$Component) {
  _inherits(DatePickerFullScreen, _React$Component);

  function DatePickerFullScreen(props) {
    _classCallCheck(this, DatePickerFullScreen);

    _get(Object.getPrototypeOf(DatePickerFullScreen.prototype), 'constructor', this).call(this, props);
    this.state = {
      currentDate: null,
      inputValue: this._getInputValueByDate(this.props.defaultDate),
      isValid: true,
      selectedDate: this.props.defaultDate,
      showCalendar: false
    };
  }

  _createClass(DatePickerFullScreen, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this = this;

      window.onkeyup = function (e) {
        if (e.keyCode === 27) {
          _this._handleCloseClick();
        }
      };
    }
  }, {
    key: '_getInputValueByDate',
    value: function _getInputValueByDate(date) {
      var inputValue = null;

      if (date) {
        var newDate = moment.unix(date);

        if (newDate.isValid()) {
          inputValue = newDate.format(this.props.format);
        } else {
          inputValue = date;
        }
      }

      return inputValue;
    }
  }, {
    key: '_getSelectedDate',
    value: function _getSelectedDate() {
      var selectedDate = this.state.selectedDate;

      return selectedDate && moment.unix(selectedDate).isValid() ? this.state.selectedDate : moment().unix();
    }
  }, {
    key: '_handleCloseClick',
    value: function _handleCloseClick() {
      this.setState({
        showCalendar: false
      });
    }
  }, {
    key: '_handleDateSelect',
    value: function _handleDateSelect(date) {
      if (this.props.closeOnDateSelect) {
        this._handleScrimClick();
      }

      this.setState({
        inputValue: moment.unix(date).format(this.props.format),
        isValid: true,
        selectedDate: date
      });

      this.props.onDateSelect(date);
    }
  }, {
    key: '_handleInputBlur',
    value: function _handleInputBlur(evt) {
      if (evt.target.value.length === 0) {
        this.props.onDateSelect(null);

        this.setState({
          inputValue: null,
          selectedDate: null
        });
      } else {
        this.setState({
          inputValue: moment.unix(this.state.selectedDate).format(this.props.format)
        });
      }
    }
  }, {
    key: '_handleInputChange',
    value: function _handleInputChange(evt) {
      this.setState({
        inputValue: evt.target.value
      });
    }
  }, {
    key: '_handlePreviousClick',
    value: function _handlePreviousClick() {
      var selectedDate = moment.unix(this._getSelectedDate()).locale(this.props.locale);
      var currentDate = this.state.currentDate ? this.state.currentDate.locale(this.props.locale) : selectedDate;

      currentDate = moment(currentDate.startOf('month').subtract(1, 'm'), this.props.format);

      this.setState({
        currentDate: currentDate
      });
    }
  }, {
    key: '_handleNextClick',
    value: function _handleNextClick() {
      var selectedDate = moment.unix(this._getSelectedDate()).locale(this.props.locale);
      var currentDate = this.state.currentDate ? this.state.currentDate.locale(this.props.locale) : selectedDate;

      currentDate = moment(currentDate.endOf('month').add(1, 'd'), this.props.format);

      this.setState({
        currentDate: currentDate
      });
    }
  }, {
    key: '_handleScrimClick',
    value: function _handleScrimClick() {
      this.setState({
        showCalendar: false
      });
    }
  }, {
    key: '_toggleCalendar',
    value: function _toggleCalendar() {
      this.setState({
        showCalendar: !this.state.showCalendar
      });
    }
  }, {
    key: '_renderMonthTable',
    value: function _renderMonthTable(currentDate, selectedDate) {
      var days = [];
      var startDate = moment(currentDate, this.props.format).startOf('month').startOf('week');
      var endDate = moment(currentDate, this.props.format).endOf('month').endOf('week');
      var minimumDate = this.props.minimumDate ? moment.unix(this.props.minimumDate) : null;

      while (startDate.isBefore(endDate)) {
        var isCurrentMonth = startDate.month() === currentDate.month();
        var isCurrentDay = startDate.format(this.props.format) === selectedDate.format(this.props.format);
        var day = undefined;
        var noSelectDay = startDate.isBefore(minimumDate);

        day = React.createElement(
          'div',
          {
            key: startDate.month() + '-' + startDate.date(),
            onClick: !noSelectDay ? this._handleDateSelect.bind(this, startDate.unix()) : null,
            style: [styles.calendarDay, !noSelectDay && isCurrentMonth && styles.currentMonth]
          },
          React.createElement(
            'div',
            {
              key: startDate.format('DDDD'),
              style: [styles.calendarDayContent, noSelectDay ? styles.calendarDayDisabled : isCurrentDay && styles.currentDay]
            },
            React.createElement(
              'div',
              { style: styles.calendarDayText },
              startDate.date()
            )
          )
        );

        if (this.props.showDayBorders) {
          day.props.style.push([styles.borderRight, styles.borderBottom]);
        }

        days.push(day);
        startDate.add(1, 'd');
      }

      return days;
    }
  }, {
    key: '_renderSelectedDate',
    value: function _renderSelectedDate() {
      if (this.props.useInputForSelectedDate) {
        var hidePlaceholder = this.state.inputValue && this.state.inputValue.length;

        return React.createElement(
          'div',
          null,
          React.createElement('input', {
            key: 'input',
            onBlur: this._handleInputBlur.bind(this),
            onChange: this._handleInputChange.bind(this),
            onClick: this._toggleCalendar.bind(this),
            style: [styles.input, this.props.inputStyle, hidePlaceholder && { backgroundColor: StyleConstants.Colors.INVERSE_PRIMARY }],
            type: 'text',
            value: this.state.inputValue
          }),
          React.createElement(
            'div',
            { style: [styles.placeholderText, this.props.placeholderTextStyle] },
            this.props.placeholderText || 'Select A Date'
          )
        );
      } else {
        return React.createElement(
          'div',
          {
            key: 'selectedDate',
            onClick: this._toggleCalendar.bind(this),
            style: styles.selectedDate
          },
          this.state.inputValue
        );
      }
    }
  }, {
    key: '_renderTitle',
    value: function _renderTitle(styles) {
      if (this.props.title) {
        return React.createElement(
          'div',
          { key: 'title', style: styles.title },
          this.props.title
        );
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var selectedDate = moment.unix(this._getSelectedDate()).locale(this.props.locale);
      var currentDate = this.state.currentDate ? this.state.currentDate.locale(this.props.locale) : selectedDate;

      return React.createElement(
        'div',
        {
          className: 'mx-date-picker-full-screen',
          style: [styles.component, styles.clearFix, this.props.style],
          tabIndex: '0'
        },
        React.createElement(
          'div',
          { className: 'mx-date-picker-full-screen-selected-date', key: 'selectedDateWrapper', style: [styles.selectedDateWrapper, this.props.selectedDateWrapperStyle] },
          this._renderSelectedDate()
        ),
        React.createElement(
          'div',
          { className: 'mx-date-picker-full-screen-calendar-scrim', key: 'calendarModal', style: [styles.calendarModal, this.state.showCalendar && styles.calendarShow, this.props.isFixed && { position: 'fixed' }] },
          React.createElement(
            'div',
            { onClick: this._handleCloseClick.bind(this), style: styles.close },
            React.createElement(Icon, {
              size: 20,
              style: styles.closeIcon,
              type: this.props.closeIcon
            }),
            React.createElement(
              'div',
              { style: styles.closeText },
              'ESC'
            )
          ),
          React.createElement(
            'div',
            { className: 'mx-date-picker-full-screen-calendar-wrapper', style: styles.calendarWrapper },
            this._renderTitle(styles),
            React.createElement(
              'div',
              { className: 'mx-date-picker-full-screen-calendar-header', key: 'calendarHeader', style: [styles.calendarHeader, { borderBottomStyle: this.props.showDayBorders ? 'solid' : 'none' }, styles.clearFix] },
              React.createElement(Icon, {
                onClick: this._handlePreviousClick.bind(this),
                size: '32px',
                style: [styles.navIcon, styles.navLeft, this.props.showDayBorders && styles.borderRight],
                type: 'caret-left'
              }),
              currentDate.format('MMMM YYYY'),
              React.createElement(Icon, {
                onClick: this._handleNextClick.bind(this),
                size: '32px',
                style: [styles.navIcon, styles.navRight, this.props.showDayBorders && styles.borderLeft],
                type: 'caret-right'
              })
            ),
            React.createElement(
              'div',
              { style: styles.calendarContainer },
              this._renderMonthTable(currentDate, selectedDate)
            ),
            React.createElement('div', { style: styles.clearFix })
          )
        )
      );
    }
  }]);

  return DatePickerFullScreen;
})(React.Component);

DatePickerFullScreen.propTypes = {
  closeIcon: React.PropTypes.string,
  closeOnDateSelect: React.PropTypes.bool,
  defaultDate: React.PropTypes.number,
  format: React.PropTypes.string,
  inputStyle: React.PropTypes.object,
  isFixed: React.PropTypes.bool,
  locale: React.PropTypes.string,
  minimumDate: React.PropTypes.number,
  onDateSelect: React.PropTypes.func,
  placeholderText: React.PropTypes.string,
  placeholderTextStyle: React.PropTypes.object,
  selectedDateWrapperStyle: React.PropTypes.object,
  showDayBorders: React.PropTypes.bool,
  style: React.PropTypes.object,
  title: React.PropTypes.string,
  useInputForSelectedDate: React.PropTypes.bool
};

DatePickerFullScreen.defaultProps = {
  closeIcon: 'close',
  closeOnDateSelect: false,
  format: 'MMM D, YYYY',
  isFixed: false,
  locale: 'en',
  onDateSelect: function onDateSelect() {},
  showDayBorders: false,
  title: 'Select A Date',
  useInputForSelectedDate: true
};

var styles = {
  calendarDay: {
    color: StyleConstants.Colors.ASH,
    float: 'left',
    paddingBottom: '11%',
    position: 'relative',
    width: '13.5%'
  },
  borderBottom: {
    borderBottom: StyleConstants.Colors.FOG,
    borderBottomStyle: 'solid',
    borderBottomWidth: 1
  },
  borderRight: {
    borderRight: StyleConstants.Colors.FOG,
    borderRightStyle: 'solid',
    borderRightWidth: 1
  },
  borderLeft: {
    borderLeft: StyleConstants.Colors.FOG,
    borderLeftStyle: 'solid',
    borderLeftWidth: 1
  },
  calendarContainer: {
    width: '100%',
    padding: '0px 2px 10px 6px'
  },
  calendarDayContent: {
    borderRadius: '50%',
    height: 32,
    left: '50%',
    position: 'absolute',
    top: '50%',
    transform: 'translateY(-50%) translateX(-50%)',
    width: 32,

    ':hover': {
      backgroundColor: StyleConstants.Colors.BLUE,
      color: StyleConstants.Colors.INVERSE_PRIMARY,
      cursor: 'pointer'
    }
  },
  calendarDayText: {
    borderRadius: '100%',
    fontSize: StyleConstants.FontSizes.MEDIUM,
    fontWeight: 'normal',
    left: '50%',
    position: 'absolute',
    top: '50%',
    transform: 'translateY(-50%) translateX(-50%)'
  },
  calendarDayDisabled: {
    ':hover': {
      background: 'none',
      color: StyleConstants.Colors.PORCELAIN
    }
  },
  calendarHeader: {
    color: StyleConstants.Colors.CHARCOAL,
    borderBottom: StyleConstants.Colors.FOG,
    borderBottomWidth: 1,
    fontSize: StyleConstants.FontSizes.XLARGE,
    fontWeight: 'normal',
    padding: '5px 0px 7px 0px',
    position: 'relative',
    textAlign: 'center',
    textTransform: 'none'
  },
  calendarIcon: {
    color: StyleConstants.Colors.PORCELAIN,
    position: 'absolute',
    right: 12.8,
    top: '50%',
    transform: 'translateY(-50%)'
  },
  calendarShow: {
    display: 'block'
  },
  close: {
    position: 'absolute',
    right: 20,
    top: 15,
    textAlign: 'center',
    cursor: 'pointer',
    color: StyleConstants.Colors.ASH
  },
  closeIcon: {
    color: StyleConstants.Colors.ASH
  },
  closeText: {
    fontSize: StyleConstants.FontSizes.TINY
  },
  clearFix: {
    clear: 'both',
    marginBottom: 15
  },
  component: {
    backgroundColor: '#fff',
    fontFamily: StyleConstants.FontFamily,
    fontSize: StyleConstants.FontSizes.LARGE,
    width: '100%',

    ':focus': {
      boxShadow: 'none',
      outline: 'none'
    }
  },
  calendarModal: {
    backgroundColor: '#fff',
    bottom: 0,
    display: 'none',
    left: 0,
    margin: 0,
    padding: 0,
    position: 'absolute',
    right: 0,
    top: 0,
    zIndex: 999
  },
  calendarWrapper: {
    height: 300,
    left: '50%',
    position: 'absolute',
    top: '50%',
    transform: 'translate(-50%, -50%)',
    width: 300
  },
  selectedDateWrapper: {
    position: 'relative',
    cursor: 'pointer'
  },
  currentDay: {
    backgroundColor: StyleConstants.Colors.BLUE,
    color: StyleConstants.Colors.INVERSE_PRIMARY
  },
  currentMonth: {
    color: StyleConstants.Colors.CHARCOAL
  },
  input: {
    backgroundColor: 'transparent',
    border: 'none',
    fontSize: StyleConstants.FontSizes.MEDIUM,
    outline: 'none',
    paddingBottom: 10,
    paddingLeft: 5,
    position: 'relative',
    top: 5,
    WebkitAppearance: 'none',
    width: '80%',
    zIndex: 2,

    ':focus': {
      border: 'none',
      boxShadow: 'none',
      outline: 'none'
    }
  },
  navIcon: {
    cursor: 'pointer'
  },
  navLeft: {
    position: 'absolute',
    left: '0',
    top: '50%',
    transform: 'translateY(-50%)'
  },
  navRight: {
    position: 'absolute',
    right: '0',
    top: '50%',
    transform: 'translateY(-50%)'
  },
  placeholderText: {
    color: StyleConstants.Colors.CHARCOAL,
    fontSize: StyleConstants.FontSizes.MEDIUM,
    paddingLeft: 5,
    position: 'absolute',
    top: 10
  },
  selectedDate: {
    color: StyleConstants.Colors.CHARCOAL,
    cursor: 'pointer',
    fontSize: StyleConstants.FontSizes.MEDIUM,
    padding: '5px 0 5px 5px',
    verticalAlign: 'middle',
    width: '100%',

    ':hover': {
      color: StyleConstants.Colors.BLUE
    }
  },
  title: {
    boxSizing: 'border-box',
    color: StyleConstants.Colors.CHARCOAL,
    fontSize: StyleConstants.FontSizes.XXLARGE,
    fontWeight: 'bold',
    padding: '0px 0px 20px 10px'
  }
};

module.exports = Radium(DatePickerFullScreen);
