// mx-react-components/dist/components/SelectFullScreen.js
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react');
var Radium = require('radium');

var Icon = require('./Icon');

var StyleConstants = require('../constants/Style');

var SelectFullScreen = (function (_React$Component) {
  _inherits(SelectFullScreen, _React$Component);

  function SelectFullScreen(props) {
    _classCallCheck(this, SelectFullScreen);

    _get(Object.getPrototypeOf(SelectFullScreen.prototype), 'constructor', this).call(this, props);
    this.state = {
      isOpen: false,
      selected: false
    };
  }

  _createClass(SelectFullScreen, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this = this;

      window.onkeyup = function (e) {
        if (e.keyCode === 27) {
          _this._handleCloseClick();
        }
      };
    }
  }, {
    key: '_handleClick',
    value: function _handleClick() {
      this.setState({
        isOpen: true
      });
    }
  }, {
    key: '_handleCloseClick',
    value: function _handleCloseClick() {
      this.setState({
        isOpen: false
      });
    }
  }, {
    key: '_handleOptionClick',
    value: function _handleOptionClick(option) {
      this.setState({
        selected: option,
        isOpen: false
      });

      this.props.onChange(option);
    }
  }, {
    key: '_handleSelectChange',
    value: function _handleSelectChange(e) {
      var selectedOption = this.props.options.filter(function (option) {
        return option.value + '' === e.target.value;
      })[0];

      this._handleOptionClick(selectedOption);
    }
  }, {
    key: '_renderOptions',
    value: function _renderOptions() {
      var _this2 = this;

      if (this.state.isOpen) {
        return React.createElement(
          'div',
          { style: [styles.optionsScrim, this.props.isFixed && { position: 'fixed' }] },
          React.createElement(
            'div',
            { onClick: this._handleCloseClick.bind(this), style: styles.close },
            React.createElement(Icon, {
              size: 20,
              style: styles.closeIcon,
              type: this.props.closeIcon
            }),
            React.createElement(
              'div',
              { style: styles.closeText },
              'ESC'
            )
          ),
          React.createElement(
            'div',
            { style: styles.content },
            React.createElement(
              'div',
              { style: styles.optionsHeader },
              this.props.optionsHeaderText
            ),
            React.createElement(
              'div',
              { className: 'mx-select-full-screen-options', style: [styles.optionsWrapper, this.props.optionsStyle] },
              this.props.options.map(function (option) {
                return React.createElement(
                  'div',
                  {
                    className: 'mx-select-full-screen-option',
                    key: option.displayValue + option.value,
                    onClick: _this2._handleOptionClick.bind(_this2, option),
                    ref: option.displayValue + option.value
                  },
                  _this2.props.optionFormatter(option)
                );
              })
            )
          )
        );
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var selected = this.state.selected || this.props.selected || { displayValue: this.props.placeholderText, value: '' };

      return React.createElement(
        'div',
        { className: 'mx-select-full-screen', style: [styles.component, this.props.style] },
        React.createElement(
          'div',
          {
            className: 'mx-select-full-screen-selected',
            key: 'selected',
            onClick: this._handleClick.bind(this),
            style: this.props.selectedStyle
          },
          selected.displayValue
        ),
        this._renderOptions()
      );
    }
  }]);

  return SelectFullScreen;
})(React.Component);

SelectFullScreen.propTypes = {
  closeIcon: React.PropTypes.string,
  isFixed: React.PropTypes.bool,
  onChange: React.PropTypes.func,
  optionFormatter: React.PropTypes.func,
  options: React.PropTypes.array,
  optionsHeaderText: React.PropTypes.string,
  optionsStyle: React.PropTypes.oneOfType([React.PropTypes.object, React.PropTypes.array]),
  optionStyle: React.PropTypes.oneOfType([React.PropTypes.object, React.PropTypes.array]),
  placeholderText: React.PropTypes.string,
  selected: React.PropTypes.object,
  selectedStyle: React.PropTypes.oneOfType([React.PropTypes.object, React.PropTypes.array])
};

SelectFullScreen.defaultProps = {
  closeIcon: 'close',
  isFixed: false,
  onChange: function onChange() {},
  optionFormatter: function optionFormatter(option) {
    return React.createElement(
      'div',
      { key: option.displayValue + option.value + '_value', style: styles.option },
      option.displayValue
    );
  },
  options: [],
  optionsHeaderText: 'Select An Option',
  placeholderText: 'Select One',
  selected: false
};

var styles = {
  close: {
    position: 'absolute',
    right: 20,
    top: 15,
    textAlign: 'center',
    cursor: 'pointer',
    color: StyleConstants.Colors.ASH
  },
  closeIcon: {
    color: StyleConstants.Colors.ASH
  },
  closeText: {
    fontSize: StyleConstants.FontSizes.TINY
  },
  component: {
    cursor: 'pointer',
    fontFamily: StyleConstants.FontFamily,
    fontSize: StyleConstants.FontSizes.LARGE,
    color: StyleConstants.Colors.CHARCOAL,
    boxSizing: 'border-box',
    outline: 'none'
  },
  content: {
    left: '50%',
    position: 'absolute',
    top: '50%',
    transform: 'translate(-50%, -50%)',
    width: 300
  },
  optionsScrim: {
    backgroundColor: '#fff',
    bottom: 0,
    height: '100%',
    left: 0,
    position: 'absolute',
    right: 0,
    top: 0,
    zIndex: 999
  },
  optionsWrapper: {
    border: '1px solid ' + StyleConstants.Colors.FOG,
    height: 250,
    overflow: 'auto',
    width: 300
  },
  option: {
    cursor: 'pointer',
    backgroundColor: '#fff',
    padding: 10,
    whiteSpace: 'nowrap',
    fontSize: StyleConstants.FontSizes.MEDIUM,

    ':hover': {
      backgroundColor: StyleConstants.Colors.PRIMARY,
      color: StyleConstants.Colors.INVERSE_PRIMARY,
      opacity: 1
    }
  },
  optionsHeader: {
    color: StyleConstants.Colors.CHARCOAL,
    fontSize: StyleConstants.FontSizes.XXLARGE,
    fontWeight: 'bold',
    paddingBottom: 10
  }
};

module.exports = Radium(SelectFullScreen);
