// mx-react-components/dist/components/TypeAhead.js
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react');
var ReactDOM = require('react-dom');
var Radium = require('radium');

var Icon = require('./Icon');

var StyleConstants = require('../constants/Style');

var TypeAhead = (function (_React$Component) {
  _inherits(TypeAhead, _React$Component);

  function TypeAhead(props) {
    _classCallCheck(this, TypeAhead);

    _get(Object.getPrototypeOf(TypeAhead.prototype), 'constructor', this).call(this, props);
    this.state = {
      highlightedValue: null,
      isOpen: false,
      searchString: '',
      selectedItems: props.preSelectedItems
    };
  }

  _createClass(TypeAhead, [{
    key: '_getFilteredItems',
    value: function _getFilteredItems() {
      var _this = this;

      return this.props.items.filter(function (item) {
        return _this.state.selectedItems.indexOf(item) === -1 && item.toLowerCase().indexOf(_this.state.searchString.toLowerCase()) > -1;
      });
    }
  }, {
    key: '_handleBlur',
    value: function _handleBlur() {
      this.setState({
        highlightedValue: null,
        isOpen: false,
        searchString: ''
      });
    }
  }, {
    key: '_handleFocus',
    value: function _handleFocus() {
      this.setState({
        isOpen: true
      });

      ReactDOM.findDOMNode(this.refs.input).focus();
    }
  }, {
    key: '_handleItemMouseOver',
    value: function _handleItemMouseOver() {
      this.setState({
        highlightedValue: null
      });
    }
  }, {
    key: '_handleSelectAll',
    value: function _handleSelectAll() {
      this.props.onItemSelect(null, this.props.items);

      this.setState({
        highlightedValue: null,
        searchString: '',
        selectedItems: this.props.items
      });
    }
  }, {
    key: '_handleClearAll',
    value: function _handleClearAll() {
      this.props.onItemSelect(null, []);

      this.setState({
        highlightedValue: null,
        searchString: '',
        selectedItems: []
      });
    }
  }, {
    key: '_handleItemSelect',
    value: function _handleItemSelect(item) {
      //add to selectedItems
      var selectedItems = this.state.selectedItems;

      selectedItems.push(item);

      this.props.onItemSelect(item, selectedItems);

      this.setState({
        highlightedValue: null,
        searchString: '',
        selectedItems: selectedItems
      });

      ReactDOM.findDOMNode(this.refs.input).focus();
    }
  }, {
    key: '_handleItemRemove',
    value: function _handleItemRemove(item) {
      var selectedItems = this.state.selectedItems.filter(function (selectedItem) {
        return selectedItem !== item;
      });

      this.props.onItemRemove(item, selectedItems);

      this.setState({
        selectedItems: selectedItems
      });

      ReactDOM.findDOMNode(this.refs.input).focus();
    }
  }, {
    key: '_handleInputKeyDown',
    value: function _handleInputKeyDown(e) {
      var searchString = e.target.value;
      var highlightedValue = this.state.highlightedValue;
      var selectedItems = this.state.selectedItems;
      var filteredItems = this._getFilteredItems();

      //add item on enter
      if (e.keyCode === 13 && highlightedValue && selectedItems.indexOf(highlightedValue) === -1) {
        this._handleItemSelect(highlightedValue);
      }

      //add first returned item on tab
      if (e.keyCode === 9) {
        e.preventDefault();

        var item = filteredItems[0];

        if (item) {
          this._handleItemSelect(item);
        }
      }

      //remove tag on backspace
      if (e.keyCode === 8 && !searchString && selectedItems.length) {
        this._handleItemRemove(selectedItems[selectedItems.length - 1]);
      }

      //highlight next item on down
      if (e.keyCode === 40) {
        e.preventDefault();
        var nextIndex = filteredItems.indexOf(highlightedValue) + 1;

        if (nextIndex < filteredItems.length) {
          this.setState({
            highlightedValue: filteredItems[nextIndex]
          });
        }

        this._scrollList(nextIndex, 'up');

        this.setState({
          selectedValue: filteredItems[nextIndex]
        });
      }

      //highlight previous item on up
      if (e.keyCode === 38) {
        e.preventDefault();
        var previousIndex = filteredItems.indexOf(highlightedValue) - 1;

        if (previousIndex > -1) {
          this.setState({
            highlightedValue: filteredItems[previousIndex]
          });
        }

        this._scrollList(previousIndex, 'down');
        this.setState({
          selectedValue: filteredItems[previousIndex]
        });
      }

      //lose foucus on esc
      if (e.keyCode === 27) {
        e.preventDefault();

        this.setState({
          searchString: '',
          isOpen: false,
          highlightedValue: null
        });

        ReactDOM.findDOMNode(this.refs.input).blur();
      }
    }
  }, {
    key: '_scrollList',
    value: function _scrollList(nextIndex, scrollDirection) {
      var filteredItems = this._getFilteredItems();
      var ul = ReactDOM.findDOMNode(this.refs.optionList);
      var skipClearSelectAll = 2;
      var activeLi = ul.children[nextIndex + skipClearSelectAll];

      if (scrollDirection === 'up' && activeLi) {
        var heightFromTop = (nextIndex + skipClearSelectAll) * activeLi.clientHeight + activeLi.clientHeight;

        if (heightFromTop > ul.clientHeight || nextIndex === 0) {
          ul.scrollTop = activeLi.offsetTop - activeLi.clientHeight * skipClearSelectAll;
        }
      } else if (scrollDirection === 'down' && activeLi) {
        var heightFromBottom = (filteredItems.length - nextIndex) * activeLi.clientHeight;

        if (heightFromBottom > ul.clientHeight) {
          ul.scrollTop = activeLi.offsetTop - activeLi.clientHeight * skipClearSelectAll;
        }

        if (nextIndex === filteredItems.length - 1) {
          ul.scrollTop = filteredItems.length * activeLi.clientHeight;
        }
      }
    }
  }, {
    key: '_handleInputChange',
    value: function _handleInputChange(e) {
      this.setState({
        searchString: e.target.value
      });
    }
  }, {
    key: '_renderSelectedItems',
    value: function _renderSelectedItems() {
      var _this2 = this;

      return this.state.selectedItems.map(function (item, index) {
        return React.createElement(
          'div',
          { className: 'mx-typeahead-selected', key: index, style: styles.itemTag },
          item,
          React.createElement(Icon, {
            onClick: _this2._handleItemRemove.bind(_this2, item),
            size: '15px',
            style: styles.removeIcon,
            type: 'close'
          })
        );
      });
    }
  }, {
    key: '_renderItemList',
    value: function _renderItemList() {
      var _this3 = this;

      return React.createElement(
        'div',
        { className: 'mx-typeahead-option-list', ref: 'optionList', style: styles.itemList },
        this.state.selectedItems.length !== this.props.items.length ? React.createElement(
          'div',
          {
            className: 'mx-typeahead-select-all',
            key: 'selectAllItem',
            onMouseDown: this._handleSelectAll.bind(this),
            onMouseOver: this._handleItemMouseOver.bind(this),
            style: styles.item
          },
          'Select All'
        ) : null,
        this.state.selectedItems.length > 0 ? React.createElement(
          'div',
          {
            className: 'mx-typeahead-clear-all',
            key: 'clearAllItem',
            onMouseDown: this._handleClearAll.bind(this),
            onMouseOver: this._handleItemMouseOver.bind(this),
            style: styles.item
          },
          'Clear'
        ) : null,
        this._getFilteredItems().map(function (item, index) {
          return React.createElement(
            'div',
            {
              className: 'mx-typeahead-option',
              key: index,
              onMouseDown: _this3._handleItemSelect.bind(_this3, item),
              onMouseOver: _this3._handleItemMouseOver.bind(_this3),
              ref: index,
              style: [styles.item, item === _this3.state.highlightedValue && styles.activeItem]
            },
            item
          );
        })
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return React.createElement(
        'div',
        {
          className: 'mx-typeahead',
          onBlur: this._handleBlur.bind(this),
          onFocus: this._handleFocus.bind(this),
          style: [styles.component, this.props.style],
          tabIndex: '0'
        },
        this._renderSelectedItems(),
        React.createElement('input', {
          className: 'mx-typeahead-input',
          key: 'input',
          onChange: this._handleInputChange.bind(this),
          onKeyDown: this._handleInputKeyDown.bind(this),
          placeholder: !this.state.selectedItems.length ? this.props.placeholderText : null,
          ref: 'input',
          style: styles.input,
          type: 'text',
          value: this.state.searchString
        }),
        React.createElement(
          'div',
          { className: 'mx-typeahead-option-list-container', style: [styles.itemListContainer, !this.state.isOpen && { display: 'none' }] },
          this._renderItemList()
        )
      );
    }
  }]);

  return TypeAhead;
})(React.Component);

var styles = {
  component: {
    backgroundColor: '#FFFFFF',
    borderColor: '#e5e5e5',
    borderRadius: '3px',
    borderStyle: 'solid',
    borderWidth: '1px',
    boxSizing: 'border-box',
    fontFamily: StyleConstants.FontFamily,
    fontSize: '12px',
    paddingTop: '10px',
    paddingRight: '10px',
    paddingBottom: '10px',
    paddingLeft: '10px',
    position: 'relative',
    WebkitAppearance: 'none',
    width: '100%',
    minHeight: '35px',

    ':focus': {
      backgroundColor: '#FFFFFF',
      boxShadow: 'none',
      color: StyleConstants.Colors.FONT,
      outline: 'none'
    }
  },
  activeItem: {
    backgroundColor: StyleConstants.Colors.PRIMARY,
    color: StyleConstants.Colors.INVERSE_PRIMARY
  },
  clearFix: {
    clear: 'both'
  },
  input: {
    backgroundColor: '#FFFFFF',
    borderWidth: 0,
    color: StyleConstants.Colors.FONT,
    fontSize: '13px',
    minWidth: '33%',
    outline: 'none',
    WebkitAppearance: 'none',

    ':focus': {
      borderWidth: 0,
      boxShadow: 'none',
      outline: 'none'
    }
  },
  itemList: {
    minHeight: '20px',
    maxHeight: '200px',
    overflow: 'auto'
  },
  itemListContainer: {
    clear: 'both',
    backgroundColor: '#fff',
    position: 'absolute',
    left: -1,
    right: -1,
    marginTop: '7px',
    marginBottom: '20px',
    border: '1px solid #E5E5E5',
    borderRadius: '0 0 3px 3px',
    boxShadow: '0 30px 30px 10px rgba(0,0,0,0.1)',
    zIndex: 10
  },
  itemTag: {
    backgroundColor: '#eee',
    borderColor: '#e5e5e5',
    borderRadius: '3px',
    borderStyle: 'solid',
    borderWidth: '1px',
    display: 'inline-block',
    lineHeight: '0.8em',
    marginTop: '1px',
    marginRight: '2px',
    marginBottom: '1px',
    paddingLeft: '3px',
    position: 'relative'
  },
  item: {
    color: StyleConstants.Colors.LIGHT_FONT,
    cursor: 'pointer',
    paddingTop: '10px',
    paddingRight: '10px',
    paddingBottom: '10px',
    paddingLeft: '10px',
    lineHeight: '1em',

    ':focus': {
      border: 'none',
      boxShadow: 'none',
      outline: 'none'
    },
    ':hover': {
      backgroundColor: StyleConstants.Colors.PRIMARY,
      color: StyleConstants.Colors.INVERSE_PRIMARY
    }
  },
  removeIcon: {
    color: StyleConstants.Colors.LIGHT_FONT,
    marginLeft: '5px',
    cursor: 'pointer'
  }
};

TypeAhead.propTypes = {
  items: React.PropTypes.array,
  onItemRemove: React.PropTypes.func,
  onItemSelect: React.PropTypes.func,
  placeholderText: React.PropTypes.string,
  preSelectedItems: React.PropTypes.array
};

TypeAhead.defaultProps = {
  items: [],
  onItemRemove: function onItemRemove() {},
  onItemSelect: function onItemSelect() {},
  placeholderText: 'Select Filters',
  preSelectedItems: []
};

module.exports = Radium(TypeAhead);
